Remove uses of chainActive and mapBlockIndex in wallet code

ryanofsky opened the pull request:
This change removes uses of `chainActive` and `mapBlockIndex` globals in wallet code. It is a refactoring change which does not affect external behavior.

This is the next step in the larger #10973 refactoring change, which removes all other accesses to node global variables from wallet code. Doing this is useful to provide a better defined interface between the wallet and node, and necessary to allow wallet and node code to run in separate processes in #10102.

ryanofsky committed:
Add height, depth, and hash methods to the Chain interface

And use them to remove uses of chainActive and mapBlockIndex in wallet code

This commit does not change behavior.

Co-authored-by: Ben Woosley <ben.woosley@gmail.com>

MarcoFalke commented:
Concept ACK. Though, I'd prefer if we first cleaned up the interface for ScanForWalletTransactions. I believe that right now it is relying on too much undefined behaviour that this refactoring could be meaningfully reviewed.

ryanofsky commented:
> too much undefined behaviour that this refactoring could be meaningfully reviewed.

Are you talking about a small part of this PR or the whole thing? Most of the changes here don't have anything to do with ScanForWalletTransactions. I could even drop the rescan changes and save them for a different PR if you are only worried about them.

practicalswift commented:
Concept ACK

Nice readability and robustness improvement

Empact commented:
Concept ACK

Would be easier/safer to review if split up. Like Marco, I’d like to see ScanFor improved first.

MeshCollider commented on src/wallet/wallet.cpp:
```diff
@@ -4122,45 +4135,55 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
-            while (block && block->pprev && (block->pprev->nStatus & BLOCK_HAVE_DATA) && block->pprev->nTx > 0 && pindexRescan != block)
-                block = block->pprev;
+            int block_height = *tip_height;
+            while (block_height > 0 && locked_chain->haveBlockOnDisk(block_height - 1) && rescan_height != block_height) {
```
Why not use `findPruned(rescan_height, *tip_height)`?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/14711#discussion_r233393592

> Why not use findPruned

I used `haveBlockOnDisk` here because I was doing a very literal translation and `findPruned` doesn't have the `block->pprev->nTx > 0` condition. If can drop that condition or add it to `findPruned`, though, if it would be an improvement.

MeshCollider commented on src/wallet/wallet.cpp:
```diff
@@ -1605,114 +1600,133 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
+ * defined by stop_block
  *
- * Caller needs to make sure pindexStop (and the optional pindexStart) are on
+ * Caller needs to make sure stop_block (and the optional start_block) are on
```
I think this is where the earlier confusion arose, it is `stop_block` which is optional not `start_block`, so the comment is wrong, the names need to be switched

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/14711#discussion_r233398045

> the names need to be switched

Good catch. I'll probably just rewrite this comment if this isn't fixed by another of the rescan PRs first. This comment never made sense to me even when it was first added in https://github.com/bitcoin/bitcoin/pull/11281#discussion_r156810727 / https://github.com/bitcoin/bitcoin/pull/11281#discussion_r163063324.

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/14711/commits/d568bdb15173dd8dd7bf95cb0bd9c9b55dbb97f0

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1526,24 +1529,18 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
 
-    const CBlockIndex* pindex = nullptr;    // Block index of the specified block or the common ancestor, if the block provided was in a deactivated chain.
-    const CBlockIndex* paltindex = nullptr; // Block index of the specified block, even if it's in a deactivated chain.
+    Optional<int> height;    // Height of the specified block or the common ancestor, if the block provided was in a deactivated chain.
```
Isn't this change conceptually different? Even though the chain is locked, this can be copied and used without the lock and therefore can "point" to other block if a reorg occurs. This wouldn't occur with `CBlockIndex`.

Looks like you could either:
 - create `interfaces::BlockIndex`
 - replace `CBlockIndex*` with `uint256

promag reviewed (commented):
Concept ACK.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -896,7 +893,7 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     void BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) override;
     int64_t RescanFromTime(int64_t startTime, const WalletRescanReserver& reserver, bool update);
-    CBlockIndex* ScanForWalletTransactions(CBlockIndex* pindexStart, CBlockIndex* pindexStop, const WalletRescanReserver& reserver, bool fUpdate = false);
+    uint256 ScanForWalletTransactions(const uint256& start_block, const uint256& stop_block, const WalletRescanReserver &reserver, bool fUpdate);
```
Haven't though about this too much, but maybe you could split out the changes in the wallet header that change the interface from taking a block index pointer to taking a block hash to a separate commit? Unless I am mistaken those type changes have the side effect of fixing the "issue" I tried to fix in #14712.

MarcoFalke commented on src/interfaces/chain.h:
```diff
@@ -38,6 +105,21 @@ class Chain
+        int64_t* max_time = nullptr) = 0;
+
+    //! Estimate fraction of total transactions verified if blocks up to
+    //! given height are verified.
```
doc-nit: height is not given (at least not directly)

ryanofsky committed:
Add time methods to the Chain interface

And use them to remove uses of chainActive and mapBlockIndex in wallet code

This commit does not change behavior.

Co-authored-by: Ben Woosley <ben.woosley@gmail.com>

ryanofsky committed:
Add findFork and findBlock to the Chain interface

And use them to remove uses of chainActive and mapBlockIndex in wallet code

This commit does not change behavior.

Co-authored-by: Ben Woosley <ben.woosley@gmail.com>

ryanofsky committed:
Convert CWallet::ScanForWalletTransactions and SyncTransaction to the new Chain apis

Only change in behavior is "Rescan started from block <height>" message
replaced by "Rescan started from block <hash>" message in
ScanForWalletTransactions.

Co-authored-by: Ben Woosley <ben.woosley@gmail.com>

ryanofsky committed:
Remove remaining chainActive references from CWallet

This commit does not change behavior.

Co-authored-by: Ben Woosley <ben.woosley@gmail.com>

Empact commented:
@ryanofsky I went ahead and split this up into a few commits, found it easier to review. Built and ran tests on each commit. https://github.com/Empact/bitcoin/tree/pr/wchain2

Empact commented on src/interfaces/chain.cpp:
```diff
@@ -4,18 +4,133 @@
 #include <validation.h>
 
 #include <memory>
+#include <unordered_map>
```
nit: I believe this belongs in `validation.h`, as `unordered_map` isn't used directly here.

Empact commented on src/interfaces/chain.cpp:
```diff
@@ -35,6 +150,34 @@ class ChainImpl : public Chain
         return std::move(result);
     }
     std::unique_ptr<Chain::Lock> assumeLocked() override { return MakeUnique<LockImpl>(); }
+    bool findBlock(const uint256& hash, CBlock* block, int64_t* time, int64_t* time_max) override
```
nit: none of the callers use more than one of the out args, so splitting this up would make the call sites simpler/clearer.

Empact commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3384,47 +3383,46 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+    uint256 failed_block, stopBlock;
     CWallet::ScanResult result =
-        pwallet->ScanForWalletTransactions(pindexStart, pindexStop, reserver, failed_block, stopBlock, true);
+        pwallet->ScanForWalletTransactions(start_block, stop_block, reserver, failed_block, stopBlock, true /* prune */);
```
`stop_block` and `stopBlock` are confusingly similar. `last_block` could be a better name for `stopBlock`.

ryanofsky commented:
Thanks @Empact! I reset the PR to your branch.

---

Split 939a400f3d632330e470c4b3e71ee822ae7be4c4 -> 7a8727256c99dfbf739448fc72082ec568e14056 ([pr/wchain2.4](https://github.com/ryanofsky/bitcoin/commits/pr/wchain2.4) -> [pr/wchain2.5](https://github.com/ryanofsky/bitcoin/commits/pr/wchain2.5)) resetting to [Empact:pr/wchain2](https://github.com/bitcoin/bitcoin/compare/master...Empact:pr/wchain2) (no diffs)

Empact commented on src/wallet/rpcdump.cpp:
```diff
@@ -366,8 +366,7 @@ UniValue importprunedfunds(const JSONRPCRequest& request)
         auto locked_chain = pwallet->chain().lock();
-        const CBlockIndex* pindex = LookupBlockIndex(merkleBlock.header.GetHash());
-        if (!pindex || !chainActive.Contains(pindex)) {
+        if (!locked_chain->getBlockHeight(merkleBlock.header.GetHash())) {
```
Could be more clear with a comparison with `nullopt`

ryanofsky reviewed (commented):
Updated ca76bcc917dbc53a1b639acbdbb5d2faf28bbc52 -> b5c5b202757e6b9322d28f4a1ac533622ab4db00 ([pr/wchain2.6](https://github.com/ryanofsky/bitcoin/commits/pr/wchain2.6) -> [pr/wchain2.7](https://github.com/ryanofsky/bitcoin/commits/pr/wchain2.7)) with suggestions from review comments
Rebased b5c5b202757e6b9322d28f4a1ac533622ab4db00 -> 5bc30a5083507fbc10691c7c26683afe7e768c88 ([pr/wchain2.7](https://github.com/ryanofsky/bitcoin/commits/pr/wchain2.7) -> [pr/wchain2.8](https://github.com/ryanofsky/bitcoin/commits/pr/wchain2.8)) due to conflict with #14556.

jamesob reviewed (approved):
utACK https://github.com/bitcoin/bitcoin/commit/b5c5b202757e6b9322d28f4a1ac533622ab4db00

My [previous comments](https://github.com/bitcoin/bitcoin/pull/10973#pullrequestreview-116397255) on a superset of this PR still hold:

> Reviewers will notice that dealing in the more primitive datatypes returned by chain methods (block heights and hashes as opposed to, say, full CBlockIndex values) results in more fiddling at callsites (null checks, arithmetic, subsequent lookups, etc.). This is an obvious drawback, but if we imagine that these calls are happening over a process boundary - as, god willing, they someday will - then the simpler the datatype the better. The stripped-down return types also more tightly accommodate the data requirements of the callers. For these two reasons, this changeset is a big step towards system decomposition IMO.

- [x] 5f71b83448 Add height, depth, and hash methods to the Chain interface
- [x] 909557ac65 Add time methods to the Chain interface
- [x] 7669e12075 Add findFork and findBlock to the Chain interface
- [x] 628f97b97e Convert CWallet::ScanForWalletTransactions and SyncTransaction to the new Chain apis

  Note for reviewers: "block" in `Chain::Lock` method names refers to a height. A height is assumed to uniquely identify a given block for the duration of Chain::Lock's lifetime.

- [x] b5c5b20275 Remove remaining chainActive references from CWallet

---

I'll do some manual testing soon.

MarcoFalke commented on src/interfaces/chain.cpp:
```diff
@@ -35,6 +150,34 @@ class ChainImpl : public Chain
+            if (it == ::mapBlockIndex.end()) {
+                return false;
+            }
+            index = it->second;
```
in commit 22abf4a61a Add findFork and findBlock to the Chain interface


Same here (use `LookupBlockIndex`)

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1622,9 +1620,9 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
     // in the blocks of the chain that was detached
     UniValue removed(UniValue::VARR);
-    while (include_removed && paltindex && paltindex != pindex) {
+    while (include_removed && altheight && *altheight > height.value_or(-1)) {
```
in commit 22abf4a61a Add findFork and findBlock to the Chain interface

Assuming that height can not be set makes no sense, since you moved the FindFork call to before the `!height` check. I'd prefer to write `*height` here.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1607,7 +1604,8 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
 
-    int depth = pindex ? (1 + chainActive.Height() - pindex->nHeight) : -1;
+    const Optional<int> tip_height = locked_chain->getHeight();
+    int depth = tip_height && height ? (1 + *tip_height - *height) : -1;
```
in commit 22abf4a Add findFork and findBlock to the Chain interface

Same (I'd prefer to not check for height again, since that has already been done above as you moved the FindFork to before the height check.)

MarcoFalke replied:
Right, my bad.

MarcoFalke commented on src/interfaces/chain.cpp:
```diff
@@ -35,6 +150,34 @@ class ChainImpl : public Chain
+    {
+        LOCK(cs_main);
+        auto it = ::mapBlockIndex.find(block_hash);
+        return GuessVerificationProgress(Params().TxData(), it != ::mapBlockIndex.end() ? it->second : nullptr);
```
in commit 65861dff7c Convert CWallet::ScanForWalletTransactions and SyncTransaction to the new Chain apis


`LookupBlockIndex` instead of access to the global mapBlockIndex?

maflcko reviewed (commented):
utACK 22abf4a61a

laanwj commented:
Some (I think) new compile warnings with this PR:
```
/.../bitcoin/src/wallet/rpcwallet.cpp: In function ‘UniValue rescanblockchain(const JSONRPCRequest&)’:
/.../bitcoin/src/wallet/rpcwallet.cpp:3408:49: warning: ‘*((void*)& tip_height +4)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
             if (*stop_height < 0 || !tip_height || *stop_height > *tip_height) {
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /.../bitcoin/src/rpc/mining.h:10:0,
                 from /.../bitcoin/src/wallet/rpcwallet.cpp:21:
/.../bitcoin/src/univalue/include/univalue.h:59:42: warning: ‘*((void*)& stop_height +4)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
     bool setInt(int val_) { return setInt((int64_t)val_); }
                                    ~~~~~~^~~~~~~~~~~~~~~
/.../bitcoin/src/wallet/rpcwallet.cpp:3391:19: note: ‘*((void*)& stop_height +4)’ was declared here
     Optional<int> stop_height;
                   ^~~~~~~~~~~
/.../bitcoin/src/wallet/rpcwallet.cpp: In function ‘UniValue listsinceblock(const JSONRPCRequest&)’:
/.../bitcoin/src/wallet/rpcwallet.cpp:1579:19: warning: ‘*((void*)& height +4)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
     Optional<int> height;    // Height of the specified block or the common ancestor, if the block provided was in a deactivated chain.
                   ^~~~~~
/.../bitcoin/src/wallet/wallet.cpp: In member function ‘CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256&, const uint256&, const WalletRescanReserver&, uint256&, uint256&, bool)’:
/.../bitcoin/src/wallet/wallet.cpp:1709:58: warning: ‘*((void*)& block_height +4)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
                         block_height = prev_block_height + 1;
                                        ~~~~~~~~~~~~~~~~~~^~~
```

ryanofsky reviewed (commented):
Updated 5bc30a5083507fbc10691c7c26683afe7e768c88 -> 1da8b4648e6b5f01a90acfe487bcda6f7a5a0424 ([pr/wchain2.8](https://github.com/ryanofsky/bitcoin/commits/pr/wchain2.8) -> [pr/wchain2.9](https://github.com/ryanofsky/bitcoin/commits/pr/wchain2.9)) to resolve comments.

---

re: https://github.com/bitcoin/bitcoin/pull/14711#issuecomment-454760017

> Some (I think) new compile warnings with this PR:

These are spurious and should be fixed if you use a newer compiler: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=47679

It might also be possible to work around them (but awkward): https://stackoverflow.com/questions/21755206/how-to-get-around-gcc-void-b-4-may-be-used-uninitialized-in-this-funct

---
re: https://github.com/bitcoin/bitcoin/pull/14711#issuecomment-454761005

> I'm also curious why LookupBlockIndex is avoided.

Wasn't really avoided, it just didn't exist when this code was written: #10973 < #11041.

jamesob commented:
re-utACK https://github.com/bitcoin/bitcoin/commit/1da8b4648e6b5f01a90acfe487bcda6f7a5a0424 per [the latest changes](https://github.com/bitcoin/bitcoin/compare/5bc30a5083507fbc10691c7c26683afe7e768c88..1da8b4648e6b5f01a90acfe487bcda6f7a5a0424)

Empact commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3387,59 +3387,58 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
-    response.pushKV("start_height", pindexStart->nHeight);
-    response.pushKV("stop_height", stopBlock->nHeight);
+    response.pushKV("start_height", start_height);
+    response.pushKV("stop_height", stop_height ? *stop_height : UniValue());
```
I think you want `stop_block` here. `stop_height` will carry the height that we sought to scan to, `stop_block` will carry the block that it actually reached.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/14711#discussion_r249200553

It would be nice to include `stop_block` in the future, since it's a hash not a height, but this is not changing the behavior at all. `stop_height` will always contain the height scanned to because if the scan fails or aborts there will have been an exception earlier.

ryanofsky reviewed (commented):
IRC mention

> [12:19:55] \<wumpus> #14711 seems almost ready for merge for a while, really just waiting for ryanofsky to respond to empact's comment on the test

http://www.erisian.com.au/bitcoin-core-dev/log-2019-01-22.html#l-146

promag commented on src/interfaces/chain.h:
```diff
@@ -38,6 +105,21 @@ class Chain
+    //! If a block pointer is provided to retrieve the block contents, and the
+    //! block exists but doesn't have data (for example due to pruning), the
+    //! block will be empty and all fields set to null.
+    virtual bool findBlock(const uint256& hash,
```
Below you call `findBlock` while there the chain is locked. So should this be in the `Chain::Lock`? Or will there be a new mutex for `mapBlockIndex`?

promag replied:
I see that the idea is to move away from locked chain so ignore my question above: https://github.com/bitcoin/bitcoin/blob/82cf6813a4ef1b4a5439eb6cddb1ab426f3c31a2/src/interfaces/chain.h#L23-L26

jnewbery reviewed (commented):
A bunch of nits, which can be ignored or addressed in a follow-up PR.

I do think you've changed the behaviour in a couple of places, notably in `rescanblockchain()`. That should be addressed before merge.

MarcoFalke commented:
Will review further after https://github.com/bitcoin/bitcoin/pull/14711#discussion_r251169892 has been fixed. Also, I clarified the documentation of the rpc in that we follow the chain in case it "runs ahead" during the rpc (https://github.com/bitcoin/bitcoin/pull/15279)

ryanofsky commented:
John's comment https://github.com/bitcoin/bitcoin/pull/14711#discussion_r251166471 made me realize there seems to be a minor race condition and inconsistent behavior in the current implementation of ScanForWalletTransactions. If there is a reorg while `cs_main` is unlocked, there are two different checks to handle the case that pindex is no longer on the new chain. One of these checks will return ScanResult::SUCCESS while the other will return ScanResult::FAILURE, and which check is hit will depend the on timing of ReadBlockFromDisk and SyncTransactions calls.

The check after the ReadBlockFromDisk call leads to ScanResult::FAILURE:

https://github.com/bitcoin/bitcoin/blob/d6e700e40f861ddd6743f4d13f0d6f6bc19093c2/src/wallet/wallet.cpp#L1677

The check after the SyncTransactions loop leads to ScanResult::SUCCESS:

https://github.com/bitcoin/bitcoin/blob/d6e700e40f861ddd6743f4d13f0d6f6bc19093c2/src/wallet/wallet.cpp#L1697
https://github.com/bitcoin/bitcoin/blob/d6e700e40f861ddd6743f4d13f0d6f6bc19093c2/src/wallet/wallet.cpp#L1664

It seems like the better thing to do would be return SUCCESS in both cases, since any blocks attached in a reorg should already be scanned for wallet transactions. Returning SUCCESS in both cases is the behavior currently implemented in this PR, but I think I should add some comments about it to clarify and maybe open a separate PR to avoid changing the behavior here.

MarcoFalke commented:
I'd prefer to keep this refactoring-only and would suggest to open a pr for that either before or after this one is merged.

ryanofsky reviewed (commented):
Updated 1da8b4648e6b5f01a90acfe487bcda6f7a5a0424 -> 44de1561aaf7556bb8ae8b582c233742ff76767d ([pr/wchain2.9](https://github.com/ryanofsky/bitcoin/commits/pr/wchain2.9) -> [pr/wchain2.10](https://github.com/ryanofsky/bitcoin/commits/pr/wchain2.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wchain2.9..pr/wchain2.10)) implementing suggestions from John.

---

re: https://github.com/bitcoin/bitcoin/pull/14711#issuecomment-458651355

> I'd prefer to keep this refactoring-only and would suggest to open a pr for that either before or after this one is merged.

Agreed. I just added a TODO referencing https://github.com/bitcoin/bitcoin/pull/14711#issuecomment-458342518. I was going to open a new PR too, but it turns out to be difficult to write a test for the new behavior, so I am putting it off for now.

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1587,132 +1587,143 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
- * @param[out] stop_block the most recent block that could be scanned,
- *     otherwise nullptr if no block could be scanned
+ * @param[in] start_block if not null, the scan will start at this block instead
+ *            of the genesis block
```
in commit db21f02648 Convert CWallet::ScanForWalletTransactions and SyncTransaction to the new Chain apis


Is this correct? It seems that without a start_block, we'd just "fall through" and return instead of starting at the genesis block.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1591,132 +1586,143 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
  * transactions for.
  */
-CWallet::ScanResult CWallet::ScanForWalletTransactions(const CBlockIndex* const pindexStart, const CBlockIndex* const pindexStop, const WalletRescanReserver& reserver, const CBlockIndex*& failed_block, const CBlockIndex*& stop_block, bool fUpdate)
+CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_block, const uint256& stop_block, const WalletRescanReserver& reserver, bool fUpdate)
```
With this rename of the second parameter (`last_block` -> `stop_block`) we now have a parameter named `stop_block` and a member in the return object called `stop_block` which refer to different things. I think this should change back to `last_block` to match the header file, and the several inline comments should be updated to say `last_block`.

EDIT: I see the RPC params are called `start_block` and `stop_block`, so this parameter naming does make sense. Could you rename the members of the return object and update the parameter names in the header file?

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1591,132 +1586,143 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
+ * @param[in] stop_block if not null, the scan will stop at this block instead
+ *            of the chain tip
  *
  * @return ScanResult indicating success or failure of the scan. SUCCESS if
```
Needs updating now that `ScanResult` is an object.

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1587,132 +1587,143 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
-            WalletLogPrintf("Rescan interrupted by shutdown request at block %d. Progress=%f\n", pindex->nHeight, progress_current);
-            return ScanResult::USER_ABORT;
+        if (block_height && fAbortRescan) {
+            WalletLogPrintf("Rescan aborted at block %d. Progress=%f\n", block_height.value_or(0), progress_current);
```
in commit db21f02648 Convert CWallet::ScanForWalletTransactions and SyncTransaction to the new Chain apis


nit: the condition checks for `block_height`, so the `value_or` is dead code and can be removed?

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1587,132 +1587,143 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
+            WalletLogPrintf("Rescan aborted at block %d. Progress=%f\n", block_height.value_or(0), progress_current);
+            result.status = ScanResult::USER_ABORT;
+        } else if (block_height && ShutdownRequested()) {
+            WalletLogPrintf("Rescan interrupted by shutdown request at block %d. Progress=%f\n", block_height.value_or(0), progress_current);
```
same

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3388,59 +3387,57 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+
+        if (tip_height) {
+            start_block = locked_chain->getBlockHash(start_height);
+            if (stop_height) {
```
A comment here could prevent other people from falling into the same trap. Something like:

```c++
// If called with a stop_height, set the stop_height here to trigger a rescan to that height.
// If called without a stop height, leave stop_height as null here so rescan continues to the tip (even if the tip advances during rescan).
```

MarcoFalke commented on src/interfaces/chain.cpp:
```diff
@@ -112,6 +117,21 @@ class LockImpl : public Chain::Lock
+        CBlockIndex* block = LookupBlockIndex(hash);
+        return block && block->GetAncestor(::chainActive.Height()) == ::chainActive.Tip();
+    }
+    CBlockLocator getLocator() override { return ::chainActive.GetLocator(); }
```
in commit 44de1561aa

Could say `getTipLocator()`?

jnewbery reviewed (commented):
(reviewing 44de1561aaf7556bb8ae8b582c233742ff76767d)

Looks great. Thanks for being so receptive to my review comments! I have just three more inline, all to do with comments and naming.

utACK 44de1561aaf7556bb8ae8b582c233742ff76767d. Comments can be cleaned up in a follow-up PR if necessary.

maflcko reviewed (approved):
utACK 44de1561aa

Couldn't find any issues with the code, just some questions on documentation and code-style.

MeshCollider commented:
Agreed, any comments/nits can be addressed in followup

ryanofsky commented:
Thanks for reviews, all! I will follow up with the remaining unaddressed comments.

The next PR in the series is #15288 (which is much simpler than this PR)

ryanofsky reviewed (commented):
Suggested changes are in #15342
